planner/core/exhaust_physical_plans.go

```diff
@@ -521,78 +538,270 @@ func (p *LogicalJoin) constructInnerIndexScan(ds *DataSource, idx *model.IndexIn
+	// This loop deal first filter out the expressions that contains columns not in index.
+	for _, innerFilter := range innerPlan.pushedDownConds {
+		affectedCols := expression.ExtractColumns(innerFilter)
+		if expression.ColumnSliceIsIntersect(affectedCols, possibleUsedKeys) {
+			remained = append(remained, innerFilter)
 			continue
 		}
-		conds = append(conds, filter)
+		rangeFilterCandidates = append(rangeFilterCandidates, innerFilter)
+	}
+	// Extract the eq/in functions of possible join key. This returned list keeps the same order with index column.
+	notKeyEqAndIn, remainedEqAndIn, rangeFilterCandidates, _ := ranger.ExtractEqAndInCondition(p.ctx, rangeFilterCandidates, notKeyIdxCols, notKeyIdxColsLen)
+	// We hope that the index cols appeared in the join keys can all be used to build range. If it cannot be satisfied,
+	// we'll mark this index as cannot be used for index join.
+	// So we should make sure that all columns before the keyMatchedLen is join key or has eq/in function.
+	if len(notKeyEqAndIn) < keyMatchedLen-matchedKeyCnt {
+		return nil, nil, nil, nil, nil
+	}
+	remained = append(remained, remainedEqAndIn...)
+	nextColPos := matchedKeyCnt + len(notKeyEqAndIn)
+	// If all cols have been considered, we can return the current result.
+	if nextColPos == len(idxCols) {
+		remained = append(remained, rangeFilterCandidates...)
+		ranges, err := p.buildTemplateRange(idxOff2keyOff, matchedKeyCnt, notKeyEqAndIn, nil, false)
+		if err != nil {
+			return nil, nil, nil, nil, err
+		}
+		return ranges, idxOff2keyOff, remained, nil, nil
+	}
+	nextCol := idxCols[nextColPos]
+	nextColCmpFilterManager := &ColWithCompareOps{
+		targetCol:         nextCol,
+		affectedColSchema: expression.NewSchema(),
+	}
+	// We first loop the other conds to see whether there's conditions can be used to build range.
+loopOtherConds:
+	for _, filter := range p.OtherConditions {
+		sf, ok := filter.(*expression.ScalarFunction)
+		if !ok || !(sf.FuncName.L == ast.LE || sf.FuncName.L == ast.LT || sf.FuncName.L == ast.GE || sf.FuncName.L == ast.GT) {
+			continue
+		}
+		if lCol, ok := sf.GetArgs()[0].(*expression.Column); ok && lCol.Equal(nil, nextCol) {
+			affectedCols := expression.ExtractColumns(sf.GetArgs()[1])
+			if len(affectedCols) == 0 {
+				continue
+			}
+			for _, col := range affectedCols {
+				if innerPlan.schema.Contains(col) {
+					continue loopOtherConds
+				}
+			}
+			nextColCmpFilterManager.appendNewExpr(sf.FuncName.L, sf.GetArgs()[1], affectedCols)
+		} else if rCol, ok := sf.GetArgs()[1].(*expression.Column); ok && rCol.Equal(nil, nextCol) {
+			affectedCols := expression.ExtractColumns(sf.GetArgs()[0])
+			if len(affectedCols) == 0 {
+				continue
+			}
+			for _, col := range affectedCols {
+				if innerPlan.schema.Contains(col) {
+					continue loopOtherConds
+				}
+			}
+			nextColCmpFilterManager.appendNewExpr(symmetricOp[sf.FuncName.L], sf.GetArgs()[0], affectedCols)
+		}
```

Comment:
Here, we can actually remove the selected filter from `OtherCond`. But since we should not change the `OtherCond` of the logical join. We need to return one more parameter

Reply:
Then will it be evaluated twice?

Reply:
Currently, yes.